File: packages/cli/src/templates/_shared/scripts/save_handoff.py
#!/usr/bin/env python3
"""Save a handoff document with folder-based sharding.

Usage:
    python .aiwcli/_shared/scripts/save_handoff.py <context_id> <<'EOF'
    # Your handoff markdown content here (with <!-- SECTION: name --> markers)
    EOF

Or with a file:
    python .aiwcli/_shared/scripts/save_handoff.py <context_id> < handoff.md

This script:
1. Parses sections from incoming markdown using <!-- SECTION: name --> markers
2. Creates a timestamped folder at _output/contexts/{context_id}/handoffs/{YYYY-MM-DD-HHMM}/
3. Writes sharded files:
   - index.md (main entry point with navigation)
   - completed-work.md, dead-ends.md, decisions.md, pending.md, context.md
   - plan.md (copy of original plan if it exists)
4. Records the event in events.jsonl (informational only)
"""
import json
import re
import subprocess
import sys
from datetime import datetime
from pathlib import Path
from typing import Dict, Optional

# Add parent directories to path for imports
SCRIPT_DIR = Path(__file__).resolve().parent
SHARED_ROOT = SCRIPT_DIR.parent
sys.path.insert(0, str(SHARED_ROOT))

from lib.context.context_manager import get_context
from lib.base.utils import eprint
from lib.base.atomic_write import atomic_write
from lib.base.constants import get_handoff_folder_path


def parse_frontmatter(content: str) -> tuple[Dict[str, str], str]:
    """Parse YAML frontmatter from markdown content.

    Returns:
        Tuple of (frontmatter dict, remaining content)
    """
    frontmatter = {}
    remaining = content

    if content.startswith('---'):
        parts = content.split('---', 2)
        if len(parts) >= 3:
            fm_lines = parts[1].strip().split('\n')
            for line in fm_lines:
                if ':' in line:
                    key, value = line.split(':', 1)
                    frontmatter[key.strip()] = value.strip()
            remaining = parts[2].strip()

    return frontmatter, remaining


def parse_handoff_sections(content: str) -> Dict[str, str]:
    """Parse markdown content by section markers.

    Looks for <!-- SECTION: name --> markers and extracts content between them.

    Returns:
        Dict mapping section names to their content
    """
    sections = {}
    current_section = None
    current_content = []

    for line in content.split('\n'):
        if line.strip().startswith('<!-- SECTION:'):
            # Save previous section
            if current_section:
                sections[current_section] = '\n'.join(current_content).strip()
            # Extract new section name
            match = re.search(r'<!-- SECTION:\s*(\S+)\s*-->', line)
            if match:
                current_section = match.group(1)
                current_content = []
        elif current_section:
            current_content.append(line)

    # Save final section
    if current_section:
        sections[current_section] = '\n'.join(current_content).strip()

    return sections


def get_git_status() -> str:
    """Get current git status."""
    try:
        result = subprocess.run(
            ['git', 'status', '--short'],
            capture_output=True,
            text=True,
            timeout=5
        )
        return result.stdout.strip() or "(no changes)"
    except Exception:
        return "(git status unavailable)"


def get_plan_path_from_context(context_id: str, project_root: Path) -> Optional[Path]:
    """Get the plan path from context.json if available."""
    context = get_context(context_id, project_root)
    if not context or not context.in_flight:
        return None

    artifact_path = context.in_flight.artifact_path
    if artifact_path:
        plan_path = Path(artifact_path)
        if plan_path.exists():
            return plan_path

    return None


def generate_index(
    frontmatter: Dict[str, str],
    sections: Dict[str, str],
    git_status: str,
    has_plan: bool,
) -> str:
    """Generate the index.md file with summary and navigation."""
    now = datetime.now()

    lines = [
        "---",
        "type: handoff",
        f"context_id: {frontmatter.get('context_id', 'unknown')}",
        f"created_at: {now.isoformat()}",
        f"session_id: {frontmatter.get('session_id', 'unknown')}",
        f"project: {frontmatter.get('project', 'unknown')}",
        f"plan_path: {frontmatter.get('plan_document', 'none')}",
        "---",
        "",
        f"# Session Handoff - {now.strftime('%Y-%m-%d %H:%M')}",
        "",
    ]

    # Summary section
    summary = sections.get('summary', '').strip()
    if summary:
        # Extract just the content (skip the ## Summary header if present)
        summary_lines = summary.split('\n')
        summary_text = '\n'.join(
            line for line in summary_lines
            if not line.strip().startswith('##')
        ).strip()
        lines.extend([
            "## Summary",
            summary_text,
            "",
        ])

    # Navigation table
    lines.extend([
        "## Quick Navigation",
        "",
        "| Document | Purpose | Priority |",
        "|----------|---------|----------|",
        "| [Dead Ends](./dead-ends.md) | Failed approaches - DO NOT RETRY | Read First |",
        "| [Pending](./pending.md) | Next steps and blockers | Action Items |",
        "| [Completed Work](./completed-work.md) | Tasks finished this session | Reference |",
        "| [Decisions](./decisions.md) | Technical choices and rationale | Reference |",
    ])

    if has_plan:
        lines.append("| [Plan](./plan.md) | Original plan being implemented | Reference |")

    lines.extend([
        "| [Context](./context.md) | External requirements and notes | Reference |",
        "",
        "## Continuation Instructions",
        "",
        "To continue this work in a new session:",
        "1. This index document provides the overview",
        "2. **Read [Dead Ends](./dead-ends.md) first** to avoid repeating failed approaches",
        "3. Check [Pending](./pending.md) for immediate next steps",
        "4. Reference other documents as needed",
        "",
        "## Git Status at Handoff",
        "```",
        git_status,
        "```",
        "",
    ])

    return '\n'.join(lines)


def write_section_file(folder: Path, filename: str, title: str, content: str) -> bool:
    """Write a section file with header."""
    lines = [
        f"# {title}",
        "",
        content if content else "(No content for this section)",
        "",
    ]

    file_path = folder / filename
    success, error = atomic_write(file_path, '\n'.join(lines))
    if not success:
        eprint(f"[save_handoff] Warning: Failed to write {filename}: {error}")
        return False
    return True


def main():
    if len(sys.argv) < 2:
        print("Usage: python save_handoff.py <context_id> < content.md", file=sys.stderr)
        print("       python save_handoff.py <context_id> <<'EOF'", file=sys.stderr)
        print("       ... markdown content with <!-- SECTION: name --> markers ...", file=sys.stderr)
        print("       EOF", file=sys.stderr)
        sys.exit(1)

    context_id = sys.argv[1]

    # Read content from stdin
    content = sys.stdin.read()
    if not content.strip():
        print("Error: No content provided via stdin", file=sys.stderr)
        sys.exit(1)

    # Project root is the parent of .aiwcli
    project_root = SHARED_ROOT.parent.parent

    # Verify context exists
    context = get_context(context_id, project_root)
    if not context:
        print(f"Error: Context not found: {context_id}", file=sys.stderr)
        sys.exit(1)

    # Parse frontmatter and sections
    frontmatter, body = parse_frontmatter(content)
    sections = parse_handoff_sections(body)

    eprint(f"[save_handoff] Parsed {len(sections)} sections: {list(sections.keys())}")

    # Create handoff folder
    handoff_folder = get_handoff_folder_path(context_id, project_root)
    handoff_folder.mkdir(parents=True, exist_ok=True)
    eprint(f"[save_handoff] Created folder: {handoff_folder}")

    # Get git status
    git_status = get_git_status()

    # Check for plan
    plan_path = get_plan_path_from_context(context_id, project_root)
    has_plan = plan_path is not None

    # Copy plan if exists
    if plan_path:
        try:
            plan_content = plan_path.read_text(encoding='utf-8')
            plan_dest = handoff_folder / "plan.md"
            success, error = atomic_write(plan_dest, plan_content)
            if success:
                eprint(f"[save_handoff] Copied plan from {plan_path}")
            else:
                eprint(f"[save_handoff] Warning: Failed to copy plan: {error}")
        except Exception as e:
            eprint(f"[save_handoff] Warning: Failed to read plan: {e}")

    # Write index.md
    index_content = generate_index(frontmatter, sections, git_status, has_plan)
    index_path = handoff_folder / "index.md"
    success, error = atomic_write(index_path, index_content)
    if not success:
        print(f"Error: Failed to write index.md: {error}", file=sys.stderr)
        sys.exit(1)

    # Write section files
    section_mapping = {
        'completed': ('completed-work.md', 'Work Completed'),
        'dead-ends': ('dead-ends.md', 'Dead Ends - Do Not Retry'),
        'decisions': ('decisions.md', 'Key Decisions'),
        'pending': ('pending.md', 'Pending Issues'),
        'next-steps': ('pending.md', None),  # Append to pending.md
        'files': ('completed-work.md', None),  # Append to completed-work.md
        'context': ('context.md', 'Context for Future Sessions'),
    }

    # Track which files we've written with their content
    file_contents: Dict[str, list] = {}

    for section_name, (filename, title) in section_mapping.items():
        section_content = sections.get(section_name, '')
        if not section_content:
            continue

        if title is None:
            # Append mode - add to existing content
            if filename not in file_contents:
                file_contents[filename] = []
            file_contents[filename].append(section_content)
        else:
            # Write mode - set as main content with title
            if filename not in file_contents:
                file_contents[filename] = [f"# {title}", "", section_content]
            else:
                # Insert title at beginning if not present
                file_contents[filename] = [f"# {title}", ""] + file_contents[filename] + ["", section_content]

    # Write all accumulated content
    for filename, content_parts in file_contents.items():
        file_path = handoff_folder / filename
        full_content = '\n'.join(content_parts) + '\n'
        success, error = atomic_write(file_path, full_content)
        if not success:
            eprint(f"[save_handoff] Warning: Failed to write {filename}: {error}")

    # Ensure all expected files exist (even if empty)
    expected_files = ['completed-work.md', 'dead-ends.md', 'decisions.md', 'pending.md', 'context.md']
    titles = {
        'completed-work.md': 'Work Completed',
        'dead-ends.md': 'Dead Ends - Do Not Retry',
        'decisions.md': 'Key Decisions',
        'pending.md': 'Pending Issues & Next Steps',
        'context.md': 'Context for Future Sessions',
    }

    for filename in expected_files:
        file_path = handoff_folder / filename
        if not file_path.exists():
            write_section_file(handoff_folder, filename, titles[filename], "")

    # Output success message (ASCII-safe for Windows)
    print(f"[OK] Created handoff folder: {handoff_folder}")
    print(f"  - index.md (entry point with navigation)")

    files_created = [f.name for f in handoff_folder.iterdir() if f.is_file() and f.name != 'index.md']
    print(f"  - {', '.join(sorted(files_created))}")

    print()
    print("Handoff document saved. Use this folder for context in the next session.")


if __name__ == "__main__":
    main()

File: packages/cli/src/templates/_shared/scripts/save_handoff.ts
#!/usr/bin/env bun
/**
 * Save a handoff document with folder-based sharding.
 *
 * Usage:
 *   bun .aiwcli/_shared/scripts/save_handoff.ts <context_id> <<'EOF'
 *   # Your handoff markdown content here (with <!-- SECTION: name --> markers)
 *   EOF
 *
 * Or with a file:
 *   bun .aiwcli/_shared/scripts/save_handoff.ts <context_id> < handoff.md
 *
 * This script:
 * 1. Parses sections from incoming markdown using <!-- SECTION: name --> markers
 * 2. Creates a timestamped folder at _output/contexts/{context_id}/handoffs/{YYYY-MM-DD-HHMM}/
 * 3. Writes sharded files:
 *    - index.md (main entry point with navigation)
 *    - completed-work.md, dead-ends.md, decisions.md, pending.md, context.md
 *    - plan.md (copy of original plan if it exists)
 * 4. Sets handoff_path and handoff_consumed=false in state.json
 */
import * as fs from "node:fs";
import * as path from "node:path";

import { getContext, saveState } from "../lib-ts/context/context-store.js";
import { getHandoffFolderPath, getProjectRoot } from "../lib-ts/base/constants.js";
import { atomicWrite } from "../lib-ts/base/atomic-write.js";
import { logInfo, logWarn, logError } from "../lib-ts/base/logger.js";
import { getGitStatusShort } from "../lib-ts/base/git-state.js";
import { eprint } from "../lib-ts/base/utils.js";

// ---------------------------------------------------------------------------
// Parsing helpers
// ---------------------------------------------------------------------------

function parseFrontmatter(content: string): [Record<string, string>, string] {
  const frontmatter: Record<string, string> = {};
  let remaining = content;

  if (content.startsWith("---")) {
    const parts = content.split("---", 3);
    if (parts.length >= 3) {
      for (const line of parts[1]!.trim().split("\n")) {
        const colonIdx = line.indexOf(":");
        if (colonIdx !== -1) {
          const key = line.slice(0, colonIdx).trim();
          const value = line.slice(colonIdx + 1).trim();
          frontmatter[key] = value;
        }
      }
      remaining = parts[2]!.trim();
    }
  }

  return [frontmatter, remaining];
}

function parseHandoffSections(content: string): Record<string, string> {
  const sections: Record<string, string> = {};
  let currentSection: string | null = null;
  const currentContent: string[] = [];

  for (const line of content.split("\n")) {
    const marker = line.trim().match(/<!-- SECTION:\s*(\S+)\s*-->/);
    if (marker) {
      if (currentSection) {
        sections[currentSection] = currentContent.join("\n").trim();
      }
      currentSection = marker[1]!;
      currentContent.length = 0;
    } else if (currentSection) {
      currentContent.push(line);
    }
  }

  if (currentSection) {
    sections[currentSection] = currentContent.join("\n").trim();
  }

  return sections;
}

// ---------------------------------------------------------------------------
// Plan helper
// ---------------------------------------------------------------------------

function getPlanPathFromContext(contextId: string, projectRoot: string): string | null {
  const context = getContext(contextId, projectRoot);
  if (!context?.plan_path) return null;
  try {
    if (fs.existsSync(context.plan_path)) return context.plan_path;
  } catch { /* ignore */ }
  return null;
}

// ---------------------------------------------------------------------------
// File generation
// ---------------------------------------------------------------------------

function generateIndex(
  frontmatter: Record<string, string>,
  sections: Record<string, string>,
  gitStatus: string,
  hasPlan: boolean,
): string {
  const now = new Date();
  const isoStr = now.toISOString();
  const dateStr = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}-${String(now.getDate()).padStart(2, "0")} ${String(now.getHours()).padStart(2, "0")}:${String(now.getMinutes()).padStart(2, "0")}`;

  const lines: string[] = [
    "---",
    "type: handoff",
    `context_id: ${frontmatter["context_id"] ?? "unknown"}`,
    `created_at: ${isoStr}`,
    `session_id: ${frontmatter["session_id"] ?? "unknown"}`,
    `project: ${frontmatter["project"] ?? "unknown"}`,
    `plan_path: ${frontmatter["plan_document"] ?? "none"}`,
    "---",
    "",
    `# Session Handoff - ${dateStr}`,
    "",
  ];

  // Summary
  const summary = (sections["summary"] ?? "").trim();
  if (summary) {
    const summaryText = summary
      .split("\n")
      .filter(l => !l.trim().startsWith("##"))
      .join("\n")
      .trim();
    lines.push("## Summary", summaryText, "");
  }

  // Navigation
  lines.push(
    "## Quick Navigation",
    "",
    "| Document | Purpose | Priority |",
    "|----------|---------|----------|",
    "| [Dead Ends](./dead-ends.md) | Failed approaches - DO NOT RETRY | Read First |",
    "| [Pending](./pending.md) | Next steps and blockers | Action Items |",
    "| [Completed Work](./completed-work.md) | Tasks finished this session | Reference |",
    "| [Decisions](./decisions.md) | Technical choices and rationale | Reference |",
  );

  if (hasPlan) {
    lines.push("| [Plan](./plan.md) | Original plan being implemented | Reference |");
  }

  lines.push(
    "| [Context](./context.md) | External requirements and notes | Reference |",
    "",
    "## Continuation Instructions",
    "",
    "To continue this work in a new session:",
    "1. This index document provides the overview",
    "2. **Read [Dead Ends](./dead-ends.md) first** to avoid repeating failed approaches",
    "3. Check [Pending](./pending.md) for immediate next steps",
    "4. Reference other documents as needed",
    "",
    "## Git Status at Handoff",
    "```",
    gitStatus,
    "```",
    "",
  );

  return lines.join("\n");
}

function writeSectionFile(folder: string, filename: string, title: string, content: string): boolean {
  const text = `# ${title}\n\n${content || "(No content for this section)"}\n`;
  const filePath = path.join(folder, filename);
  const [success, error] = atomicWrite(filePath, text);
  if (!success) {
    logWarn("save_handoff", `Failed to write ${filename}: ${error}`);
    return false;
  }
  return true;
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

function main(): void {
  if (process.argv.length < 3) {
    eprint(
      "Usage: bun save_handoff.ts <context_id> < content.md\n" +
      "       bun save_handoff.ts <context_id> <<'EOF'\n" +
      "       ... markdown content with <!-- SECTION: name --> markers ...\n" +
      "       EOF",
    );
    process.exit(1);
  }

  const contextId = process.argv[2]!;

  // Read content from stdin
  let content: string;
  try {
    content = fs.readFileSync(0, "utf-8");
  } catch {
    logError("save_handoff", "Failed to read from stdin");
    process.exit(1);
    return; // unreachable but makes TS happy
  }

  if (!content.trim()) {
    logError("save_handoff", "No content provided via stdin");
    process.exit(1);
  }

  // Project root via shared utility (checks CLAUDE_PROJECT_DIR, falls back to cwd)
  const projectRoot = getProjectRoot(process.cwd());

  // Verify context exists
  const context = getContext(contextId, projectRoot);
  if (!context) {
    logError("save_handoff", `Context not found: ${contextId}`);
    process.exit(1);
  }

  // Parse frontmatter and sections
  const [frontmatter, body] = parseFrontmatter(content);
  const sections = parseHandoffSections(body);

  logInfo("save_handoff", `Parsed ${Object.keys(sections).length} sections: ${Object.keys(sections).join(", ")}`);

  // Create handoff folder
  const handoffFolder = getHandoffFolderPath(contextId, projectRoot);
  fs.mkdirSync(handoffFolder, { recursive: true });
  logInfo("save_handoff", `Created folder: ${handoffFolder}`);

  // Git status
  const gitStatus = getGitStatusShort(projectRoot);

  // Check for plan
  const planPath = getPlanPathFromContext(contextId, projectRoot);
  const hasPlan = planPath !== null;

  // Copy plan if exists
  if (planPath) {
    try {
      const planContent = fs.readFileSync(planPath, "utf-8");
      const [success, error] = atomicWrite(path.join(handoffFolder, "plan.md"), planContent);
      if (success) {
        logInfo("save_handoff", `Copied plan from ${planPath}`);
      } else {
        logWarn("save_handoff", `Failed to copy plan: ${error}`);
      }
    } catch (e) {
      logWarn("save_handoff", `Failed to read plan: ${e}`);
    }
  }

  // Write index.md
  const indexContent = generateIndex(frontmatter, sections, gitStatus, hasPlan);
  const indexPath = path.join(handoffFolder, "index.md");
  {
    const [success, error] = atomicWrite(indexPath, indexContent);
    if (!success) {
      logError("save_handoff", `Failed to write index.md: ${error}`);
      process.exit(1);
    }
  }

  // Write section files
  const sectionMapping: Record<string, [string, string | null]> = {
    completed: ["completed-work.md", "Work Completed"],
    "dead-ends": ["dead-ends.md", "Dead Ends - Do Not Retry"],
    decisions: ["decisions.md", "Key Decisions"],
    pending: ["pending.md", "Pending Issues"],
    "next-steps": ["pending.md", null],        // Append to pending.md
    files: ["completed-work.md", null],         // Append to completed-work.md
    context: ["context.md", "Context for Future Sessions"],
  };

  // Track accumulated content per file
  const fileContents: Record<string, string[]> = {};

  for (const [sectionName, [filename, title]] of Object.entries(sectionMapping)) {
    const sectionContent = sections[sectionName];
    if (!sectionContent) continue;

    if (title === null) {
      // Append mode
      if (!fileContents[filename]) fileContents[filename] = [];
      fileContents[filename]!.push(sectionContent);
    } else {
      // Write mode with title
      if (!fileContents[filename]) {
        fileContents[filename] = [`# ${title}`, "", sectionContent];
      } else {
        fileContents[filename] = [`# ${title}`, "", ...fileContents[filename]!, "", sectionContent];
      }
    }
  }

  // Write all accumulated content
  for (const [filename, parts] of Object.entries(fileContents)) {
    const filePath = path.join(handoffFolder, filename);
    const [success, error] = atomicWrite(filePath, parts.join("\n") + "\n");
    if (!success) {
      logWarn("save_handoff", `Failed to write ${filename}: ${error}`);
    }
  }

  // Ensure all expected files exist (even if empty)
  const expectedFiles: Record<string, string> = {
    "completed-work.md": "Work Completed",
    "dead-ends.md": "Dead Ends - Do Not Retry",
    "decisions.md": "Key Decisions",
    "pending.md": "Pending Issues & Next Steps",
    "context.md": "Context for Future Sessions",
  };

  for (const [filename, title] of Object.entries(expectedFiles)) {
    const filePath = path.join(handoffFolder, filename);
    if (!fs.existsSync(filePath)) {
      writeSectionFile(handoffFolder, filename, title, "");
    }
  }

  // Set handoff_path and handoff_consumed=false in state.json
  try {
    const indexPathStr = path.join(handoffFolder, "index.md");
    const state = getContext(contextId, projectRoot);
    if (state) {
      state.handoff_path = indexPathStr;
      state.handoff_consumed = false;
      const [ok, err] = saveState(contextId, state, projectRoot);
      if (ok) {
        logInfo("save_handoff", `Set handoff_path: ${indexPathStr}`);
      } else {
        logWarn("save_handoff", `Failed to save state: ${err}`);
      }
    } else {
      logWarn("save_handoff", `Could not load context state for ${contextId}`);
    }
  } catch (e) {
    logWarn("save_handoff", `Handoff saved but auto-resume won't work (context update failed): ${e}`);
  }

  // Output success message
  console.log(`[OK] Created handoff folder: ${handoffFolder}`);
  console.log("  - index.md (entry point with navigation)");

  const filesCreated = fs.readdirSync(handoffFolder)
    .filter(f => f !== "index.md" && fs.statSync(path.join(handoffFolder, f)).isFile())
    .sort();
  console.log(`  - ${filesCreated.join(", ")}`);

  console.log("");
  console.log("Handoff document saved. Use this folder for context in the next session.");
}

main();
